## Naive Bayes on Political Text

In this notebook we use Naive Bayes to explore and classify political data. See the `README.md` for full details. You can download the required DB from the shared dropbox or from blackboard

In [1]:
import sqlite3
import nltk
import random
import re
import numpy as np
from collections import Counter, defaultdict
from nltk.corpus import stopwords
import string
from string import punctuation

# Feel free to include your text patterns functions
#from text_functions_solutions import clean_tokenize, get_patterns

In [2]:
convention_db = sqlite3.connect("/Users/kyledalope/Downloads/nb-assignment-data/2020_Conventions.db") #connects to db
convention_cur = convention_db.cursor() #cursor object from db, to execute queries

### Part 1: Exploratory Naive Bayes

We'll first build a NB model on the convention data itself, as a way to understand what words distinguish between the two parties. This is analogous to what we did in the "Comparing Groups" class work. First, pull in the text 
for each party and prepare it for use in Naive Bayes.  

In [3]:
#Identify table names 

convention_cur.execute("SELECT name FROM sqlite_master WHERE type='table';")

tables = convention_cur.fetchall()

# Check if there are any tables
if len(tables) == 0:
    print("No tables found in the database.")
else:
    # Print the table names
    for table in tables:
        print(table[0])  

conventions


In [4]:
#identify column names of conventions table

convention_cur.execute("PRAGMA table_info(conventions)")
columns = convention_cur.fetchall()

# Check if there are any columns
if len(columns) == 0:
    print("No columns found in the table.")
else:
    # Print the column names
    for column in columns:
        print(column[1])


party
night
speaker
speaker_count
time
text
text_len
file


In [5]:
convention_data = []

# fill this list up with items that are themselves lists. The 
# first element in the sublist should be the cleaned and tokenized
# text in a single string. The second element should be the party. 

punctuation = set(punctuation) # speeds up comparison
tw_punct = punctuation - {"#"}

# Stopwords
sw = stopwords.words("english")

# Two useful regex
#whitespace_pattern = re.compile(r"\s+")
#hashtag_pattern = re.compile(r"^#[0-9a-zA-Z]+")

def remove_stop(tokens) :
    return [token for token in tokens if token.lower() not in sw]

def remove_punctuation(text, punct_set=tw_punct) :
    return("".join([ch for ch in text if ch not in punct_set]))

def tokenize(text) :
    """ Splitting on whitespace rather than the book's tokenize function. That
    function will drop tokens like '#hashtag' or '2A', which we need for Twitter. """
    return(text.split())

def prepare(text, pipeline) :
    tokens = str(text)

    for transform in pipeline :
        tokens = transform(tokens)
    
    return(tokens)

query_results = convention_cur.execute(
                            '''
                            SELECT text, party FROM conventions
                            ''')

for row in query_results :
    # store the results in convention_data
    text = row[0]
    party = row[1]

    #clean and tokenize text column
    text_clean = text
    #text_clean = whitespace_pattern.sub(" ", text_clean)
    text_clean = remove_punctuation(text_clean)
    text_clean = tokenize(text_clean)
    text_clean = remove_stop(text_clean)
    
    #append cleaned text data to party column
    convention_data.append([text_clean, party])
    
    
    #pass # remove this
    


Let's look at some random entries and see if they look right. 

In [6]:
random.choices(convention_data,k=10)

[[['there’s',
   'Pastor',
   'Aaron',
   'Johnson',
   'Today',
   'he’s',
   'executive',
   'director',
   'Tulsa',
   'Dream',
   'Center',
   'nonprofit',
   'provides',
   'education',
   'medical',
   'services',
   'food',
   'need'],
  'Republican'],
 [['Life',
   'extremely',
   'dangerous',
   'growing',
   'John',
   'Lewis',
   'respect',
   'everybody',
   'one',
   'demonstrated',
   'courage',
   'He’d',
   'beaten',
   'knocked',
   'get',
   'go',
   'fight',
   'another',
   'battle',
   'John',
   'focused',
   'ending',
   'voter',
   'suppression',
   'wasn’t',
   'great',
   'orator',
   'great',
   'spirit',
   'power',
   'spirituality',
   'humility',
   'willingness',
   'suffer',
   'rather',
   'inflict',
   'suffering'],
  'Democratic'],
 [['one',
   'qualified',
   'Joe',
   'Biden',
   'sitting',
   'head',
   'table',
   'Situation',
   'Room',
   'making',
   'decisions',
   'country'],
  'Democratic'],
 [['Make',
   'mistake',
   'didn’t',
   'bad',
 

If that looks good, we now need to make our function to turn these into features. In my solution, I wanted to keep the number of features reasonable, so I only used words that occur at least `word_cutoff` times. Here's the code to test that if you want it. 

In [7]:
word_cutoff = 5

tokens = [w for t, p in convention_data for w in prepare([t], [tokenize, remove_stop])]

word_dist = nltk.FreqDist(tokens)

feature_words = set()

for word, count in word_dist.items() :
    if count > word_cutoff :
        feature_words.add(word)
        
print(f"With a word cutoff of {word_cutoff}, we have {len(feature_words)} as features in the model.")

With a word cutoff of 5, we have 2505 as features in the model.


In [8]:
def conv_features(text,fw) :
    """Given some text, this returns a dictionary holding the
       feature words.
       
       Args: 
            * text: a piece of text in a continuous string. Assumes
            text has been cleaned and case folded.
            * fw: the *feature words* that we're considering. A word 
            in `text` must be in fw in order to be returned. This 
            prevents us from considering very rarely occurring words.
        
       Returns: 
            A dictionary with the words in `text` that appear in `fw`. 
            Words are only counted once. 
            If `text` were "quick quick brown fox" and `fw` = {'quick','fox','jumps'},
            then this would return a dictionary of 
            {'quick' : True,
             'fox' :    True}
        
    """
    
    # Your code here

    ret_dict = dict()
    clean_fw = set(word.lower().strip(string.punctuation) for word in fw)
    for word in text.lower().split():
        if word in clean_fw:
            ret_dict[word] = True

    return ret_dict


In [9]:
assert(len(feature_words)>0)
assert(conv_features("donald is the president",feature_words)==
       {'donald':True,'president':True})
assert(conv_features("people are american in america",feature_words)==
                     {'america':True,'american':True,"people":True})

Now we'll build our feature set. Out of curiosity I did a train/test split to see how accurate the classifier was, but we don't strictly need to since this analysis is exploratory. 

In [10]:
featuresets = [(conv_features(" ".join(text), feature_words), party) for (text, party) in convention_data]

In [11]:
#featuresets = [(conv_features(text_clean,feature_words), party) for (text_clean, party) in convention_data]

In [12]:
random.seed(20220507)
random.shuffle(featuresets)

test_size = 500

In [13]:
test_set, train_set = featuresets[:test_size], featuresets[test_size:]
classifier = nltk.NaiveBayesClassifier.train(train_set)
print(nltk.classify.accuracy(classifier, test_set))

0.494


In [14]:
classifier.show_most_informative_features(25)

Most Informative Features
                   china = True           Republ : Democr =     25.8 : 1.0
                   votes = True           Democr : Republ =     23.8 : 1.0
             enforcement = True           Republ : Democr =     21.5 : 1.0
                 destroy = True           Republ : Democr =     19.2 : 1.0
                freedoms = True           Republ : Democr =     18.2 : 1.0
                 climate = True           Democr : Republ =     17.8 : 1.0
                supports = True           Republ : Democr =     17.1 : 1.0
                   crime = True           Republ : Democr =     16.1 : 1.0
                   media = True           Republ : Democr =     14.9 : 1.0
                 beliefs = True           Republ : Democr =     13.0 : 1.0
               countries = True           Republ : Democr =     13.0 : 1.0
                 defense = True           Republ : Democr =     13.0 : 1.0
                    isis = True           Republ : Democr =     13.0 : 1.0

Write a little prose here about what you see in the classifier. Anything odd or interesting?

### My Observations

Based on the "most informative features" provided, most of the rhetoric in the text data is related to the Republican party. The few that were more favored to be Democratic was only 'votes' and 'climate'. While everything else was interestingly Republican; however, while keeping up with politics that seems to be fairly accurate especially with the top 5 words being 'china', 'votes, 'enforcement', 'destroy', and 'freedoms'. I believe those words were always being stated by politicians or by the media for each party. 



## Part 2: Classifying Congressional Tweets

In this part we apply the classifer we just built to a set of tweets by people running for congress
in 2018. These tweets are stored in the database `congressional_data.db`. That DB is funky, so I'll
give you the query I used to pull out the tweets. Note that this DB has some big tables and 
is unindexed, so the query takes a minute or two to run on my machine.

In [15]:
cong_db = sqlite3.connect("/Users/kyledalope/Downloads/nb-assignment-data/congressional_data.db")
cong_cur = cong_db.cursor()

In [21]:
results = cong_cur.execute(
        '''
           SELECT DISTINCT 
                  cd.candidate, 
                  cd.party,
                  tw.tweet_text
           FROM candidate_data cd 
           INNER JOIN tweets tw ON cd.twitter_handle = tw.handle 
               AND cd.candidate == tw.candidate 
               AND cd.district == tw.district
           WHERE cd.party in ('Republican','Democratic') 
               AND tw.tweet_text NOT LIKE '%RT%'
        ''')

results = list(results) # Just to store it, since the query is time consuming

In [22]:
results #observe tweet data

[('Mo Brooks',
  'Republican',
  b'"Brooks Joins Alabama Delegation in Voting Against Flawed Funding Bill" http://t.co/3CwjIWYsNq'),
 ('Mo Brooks',
  'Republican',
  b'"Brooks: Senate Democrats Allowing President to Give Americans\xe2\x80\x99 Jobs to Illegals" #securetheborder https://t.co/mZtEaX8xS6'),
 ('Mo Brooks',
  'Republican',
  b'"NASA on the Square" event this Sat. 11AM \xe2\x80\x93 4PM. Stop by &amp; hear about the incredible work done in #AL05! @DowntownHSV http://t.co/R9zY8WMEpA'),
 ('Mo Brooks',
  'Republican',
  b'"The trouble with Socialism is that eventually you run out of other people\'s money." - Margaret Thatcher https://t.co/X97g7wzQwJ'),
 ('Mo Brooks',
  'Republican',
  b'"The trouble with socialism is eventually you run out of other people\'s money" \xe2\x80\x93 Thatcher. She\'ll be sorely missed. http://t.co/Z8gBnDQUh8'),
 ('Mo Brooks',
  'Republican',
  b'"U.S. Debt Biggest Threat to National Security" #armyaviation http://t.co/4RWjrehH'),
 ('Mo Brooks',
  'Repu

In [18]:
punctuation = set(punctuation) # speeds up comparison
tw_punct = punctuation - {"#"}

# Stopwords
sw = stopwords.words("english")

# Two useful regex
#whitespace_pattern = re.compile(r"\s+")
#hashtag_pattern = re.compile(r"^#[0-9a-zA-Z]+")

def remove_stop(tokens) :
    return [token for token in tokens if token.lower() not in sw]

def remove_punctuation(text, punct_set=tw_punct) :
    return("".join([ch for ch in text if ch not in punct_set]))

def tokenize(text) :
    """ Splitting on whitespace rather than the book's tokenize function. That
    function will drop tokens like '#hashtag' or '2A', which we need for Twitter. """
    return(text.split())

def prepare(text, pipeline) :
    tokens = str(text)

    for transform in pipeline :
        tokens = transform(tokens)
    
    return(tokens)

tweet_data = []

# Now fill up tweet_data with sublists like we did on the convention speeches.
# Note that this may take a bit of time, since we have a lot of tweets.

for row in results:
    candidate = row[0]
    party = row[1]
    tweet_text = str(row[2])

    # Preprocess tweet text
    tweet_clean = remove_punctuation(tweet_text)
    tweet_clean = tokenize(tweet_clean)
    tweet_clean = remove_stop(tweet_clean)

    # Append processed tweet data to tweet_data list
    tweet_data.append([tweet_clean, party])

There are a lot of tweets here. Let's take a random sample and see how our classifer does. I'm guessing it won't be too great given the performance on the convention speeches...

In [23]:
random.seed(20201014)

tweet_data_sample = random.choices(tweet_data,k=10)

In [28]:

for tweet, party in tweet_data_sample :
    tweet_text
    estimated_party = classifier.classify({token: True for token in tweet}) #used from part 1
    
    # Fill in the right-hand side above with code that estimates the actual party
    
    print(f"Here's our (cleaned) tweet: {tweet}")
    print(f"Actual party is {party} and our classifer says {estimated_party}.")
    print("")
    

Here's our (cleaned) tweet: ['bEarlier', 'today', 'spoke', 'House', 'Floor', 'abt', 'protecting', 'health', 'care', 'women', 'praised', 'PPmarmonte', 'work', 'Central', 'Coast', 'httpstcoWqgTRzT7VV']
Actual party is Democratic and our classifer says Republican.

Here's our (cleaned) tweet: ['bGo', 'Tribe', '#RallyTogether', 'httpstco0NXutFL9L5']
Actual party is Democratic and our classifer says Democratic.

Here's our (cleaned) tweet: ['bApparently', 'Trump', 'thinks', 'easy', 'students', 'overwhelmed', 'crushing', 'burden', 'debt', 'pay', 'student', 'loans', '#TrumpBudget', 'httpstcockYQO5T0Qh']
Actual party is Democratic and our classifer says Republican.

Here's our (cleaned) tweet: ['bWexe2x80x99re', 'grateful', 'first', 'responders', 'rescue', 'personnel', 'firefighters', 'police', 'volunteers', 'working', 'tirelessly', 'keep', 'people', 'safe', 'provide', 'muchneeded', 'help', 'putting', 'lives', 'linennhttpstcoeZPv0vMIz3']
Actual party is Republican and our classifer says Republ

Now that we've looked at it some, let's score a bunch and see how we're doing.

In [31]:
# dictionary of counts by actual party and estimated party. 
# first key is actual, second is estimated
parties = ['Republican','Democratic']
results = defaultdict(lambda: defaultdict(int))

for p in parties :
    for p1 in parties :
        results[p][p1] = 0


num_to_score = 10000 #number of tweets to be scored
random.shuffle(tweet_data)

for idx, tp in enumerate(tweet_data) :
    tweet, party = tp    
    # Now do the same thing as above, but we store the results rather
    # than printing them. 
   
    # get the estimated party
    estimated_party = classifier.classify({token: True for token in tweet})
    
    results[party][estimated_party] += 1
    
    if idx > num_to_score : 
        break

In [32]:
results

defaultdict(<function __main__.<lambda>()>,
            {'Republican': defaultdict(int,
                         {'Republican': 3202, 'Democratic': 1170}),
             'Democratic': defaultdict(int,
                         {'Republican': 4234, 'Democratic': 1396})})

### Reflections

Based on the results, for the Republican tweets the NB classifier was able to classify 3202 as Republican and 1170 as Democractic. As for Democratic tweets, it classified 4234 as Republican and 1396 as Democractic. It seems the classification was not that accurate and heavily led towards Republican as the classifier in the results. I would further analyze the trainining data and see if further pre-processing steps are required to increase the accuracy. For example, in the pre-processing steps I could consider the use of regular expressions for text processing a bit better or to handle a class imbalance of the party column. However, the Naive Bayes classifier does tend to perform less when compared to other classifiers such as SVM, K-NN, or random forest. 

# References

How do I trim whitespace? (n.d.). Stack Overflow. Retrieved June 3, 2023, from https://stackoverflow.com/questions/1185524/how-do-i-trim-whitespace

‌sqlite3 — DB-API 2.0 interface for SQLite databases — Python 3.8.2 documentation. (n.d.). Docs.python.org. https://docs.python.org/3/library/sqlite3.html

‌5. Data Structures — Python 3.8.3 documentation. (n.d.). Docs.python.org. https://docs.python.org/3/tutorial/datastructures.html

‌How To Classify Data In Python using Scikit-learn. (n.d.). ActiveState. https://www.activestate.com/resources/quick-reads/how-to-classify-data-in-python/

‌